File: basic_dsa_forty.c
/**
 * Basic DSA Problems - Binary Search (C)
 * Find index of target in sorted array, or -1 if not present
 */
#include <stdio.h>

/**
 * Classic binary search: narrow [lo, hi] by comparing mid with target.
 * Time O(log n), Space O(1)
 */
int binary_search(int *nums, int n, int target) {
    int lo = 0, hi = n - 1;
    while (lo <= hi) {
        int mid = lo + (hi - lo) / 2;
        if (nums[mid] == target)
            return mid;
        if (nums[mid] < target)
            lo = mid + 1;
        else
            hi = mid - 1;
    }
    return -1;
}

int main(void) {
    printf("============================================================\n");
    printf("Problem: Binary Search\n");
    printf("============================================================\n");

    int nums[] = {-1, 0, 3, 5, 9, 12};
    int n = sizeof(nums) / sizeof(nums[0]);

    printf("target 9 -> %d\n", binary_search(nums, n, 9));
    printf("target 2 -> %d\n", binary_search(nums, n, 2));

    return 0;
}

File: basic_dsa_forty.py
"""
Basic DSA - Maximum Subarray Sum (Kadane's Algorithm)
Find the contiguous subarray with the largest sum.
"""


def max_subarray_sum(nums):
    """
    Kadane's algorithm.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if not nums:
        return 0

    best = current = nums[0]
    for x in nums[1:]:
        current = max(x, current + x)
        best = max(best, current)
    return best


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Maximum Subarray Sum (Kadane's Algorithm)")
    print("=" * 60)

    tests = [
        ([-2, 1, -3, 4, -1, 2, 1, -5, 4], 6),
        ([1], 1),
        ([5, 4, -1, 7, 8], 23),
        ([-1, -2, -3], -1),
        ([], 0),
    ]

    for nums, expected in tests:
        got = max_subarray_sum(nums)
        print(f"nums={nums} -> {got} (expected {expected})")

